### PR TITLE
Add shared examples for specifying rspecs

### PR DESCRIPTION
**Pull request by nepalez**
After loading `transproc/rspec` two set of shared examples made available:
* `:transforming_immutable_data` checks that calling `described_class.[]` with
  given arguments transforms immutable `input` into the `output`.
* `:mutating_input_data` checks that calling `decribed_class.[]` with
  given arguments mutates the `input` into the `output`

Example:
```ruby
  require 'transproc/rspec'

  describe Transproc::Coersions do # described_class
    it_behaves_like :transforming_immutable_data do
      let(:arguments) { [:to_string] }
      let(:input)  { :foo  }
      let(:output) { "foo" }
    end
  end
```

Personally, I heavily use this examples (tbh, mainly immutable) in a new faceter. I define them in `abstract_mapper`, but think the `transproc` is a better place.

PS. I rewrote a couple of specs (`fold` and `fold!`) in this style just for example.

### CODE REVIEW

File: lib/transproc/rspec.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# Examples for testing transproc functions
+# ==============================================================================
+
+shared_examples :transforming_immutable_data do
+
+  let(:fn) { described_class[*arguments] }
+
+  let(:immutable_input) do
+    begin
+      input.dup.freeze
```
Comment (solnic) [inline lib/transproc/rspec.rb]: we should use `ice_nine` and deep freeze it

**Comment by solnic** [comment]
Yeah that's cool. Maybe we should also add an expectation that original input is still the same after calling a function.

**Comment by nepalez** [comment]
@solnic `:transforming_immutable_data` duplicates and freezes the input when this is possible. So immutability of input is guaranteed.

File: lib/transproc/rspec.rb (hunk 1 above)
Comment (nepalez) [inline lib/transproc/rspec.rb]: wow! thanks, i didn't known about this gem

**Comment by solnic** [comment]
So it crashes rather than giving a spec failure message?

**Comment by nepalez** [comment]
Yes... well, you're right, possibly not freezing, but checking the input is a better solution

**Comment by nepalez** [comment]
@solnic Updated

**Pull request merged**